// lru-memoizer/lib/async.js
"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var lru_cache_1 = __importDefault(require("lru-cache"));
var events_1 = require("events");
var lodash_clonedeep_1 = __importDefault(require("lodash.clonedeep"));
var freeze_1 = require("./freeze");
var sync_1 = require("./sync");
function asyncMemoizer(options) {
    var cache = new lru_cache_1.default(options);
    var load = options.load;
    var hash = options.hash;
    var bypass = options.bypass;
    var itemMaxAge = options.itemMaxAge;
    var freeze = options.freeze;
    var clone = options.clone;
    var loading = new Map();
    var emitter = new events_1.EventEmitter();
    var defaultResult = Object.assign({
        del: del,
        reset: function () { return cache.reset(); },
        keys: cache.keys.bind(cache),
        on: emitter.on.bind(emitter),
        once: emitter.once.bind(emitter)
    }, options);
    if (options.disable) {
        return Object.assign(load, defaultResult);
    }
    function del() {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        var key = hash.apply(void 0, __spread(args));
        cache.del(key);
    }
    function emit(event) {
        var parameters = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            parameters[_i - 1] = arguments[_i];
        }
        emitter.emit.apply(emitter, __spread([event], parameters));
    }
    var result = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        var parameters = args.slice(0, -1);
        var callback = args.slice(-1).pop();
        var key;
        if (bypass && bypass.apply(void 0, __spread(parameters))) {
            emit.apply(void 0, __spread(['miss'], parameters));
            return load.apply(void 0, __spread(args));
        }
        if (parameters.length === 0 && !hash) {
            //the load function only receives callback.
            key = '_';
        }
        else {
            key = hash.apply(void 0, __spread(parameters));
        }
        var fromCache = cache.get(key);
        if (fromCache) {
            emit.apply(void 0, __spread(['hit'], parameters));
            if (clone) {
                return callback.apply(void 0, __spread([null].concat(fromCache).map(lodash_clonedeep_1.default)));
            }
            return callback.apply(void 0, __spread([null].concat(fromCache)));
        }
        if (!loading.get(key)) {
            emit.apply(void 0, __spread(['miss'], parameters));
            loading.set(key, []);
            // @ts-ignore
            load.apply(void 0, __spread(parameters.concat(function () {
                var args = [];
                for (var _i = 0; _i < arguments.length; _i++) {
                    args[_i] = arguments[_i];
                }
                var err = args[0];
                //we store the result only if the load didn't fail.
                if (!err) {
                    var result_1 = args.slice(1);
                    if (freeze) {
                        args.forEach(freeze_1.deepFreeze);
                    }
                    if (itemMaxAge) {
                        // @ts-ignore
                        cache.set(key, result_1, itemMaxAge.apply(void 0, __spread(parameters.concat(result_1))));
                    }
                    else {
                        cache.set(key, result_1);
                    }
                }
                //immediately call every other callback waiting
                var waiting = loading.get(key).concat(callback);
                loading.delete(key);
                waiting.forEach(function (callback) {
                    if (clone) {
                        return callback.apply(void 0, __spread(args.map(lodash_clonedeep_1.default)));
                    }
                    callback.apply(void 0, __spread(args));
                });
                /////////
            })));
        }
        else {
            emit.apply(void 0, __spread(['queue'], parameters));
            loading.get(key).push(callback);
        }
    };
    // @ts-ignore
    return Object.assign(result, defaultResult);
}
exports.asyncMemoizer = asyncMemoizer;
asyncMemoizer.sync = sync_1.syncMemoizer;
